addons/web_editor/static/src/js/wysiwyg/widgets/link_dialog.js

```diff
@@ -184,6 +185,11 @@ var LinkDialog = Dialog.extend({
 
         if (url.indexOf('@') >= 0 && url.indexOf('mailto:') < 0 && !url.match(/^http[s]?/i)) {
             url = ('mailto:' + url);
+        } else if (url.indexOf(location.origin) === 0 && this.$('#o_link_dialog_url_strip_domain').prop("checked")) {
+            url = url.slice(location.origin.length);
+            if (url.indexOf(location.pathname + "#") === 0) {
```

Comment:
Imo, this if is useless.

I understand what it does, but doesn't see the interest.
And even without it, your initial problem should be fixed.

Reply:
I thought about it, but there's still the case of page renames.

However it's true that for menu items it would be a problem instead. Let me remove it.